doc/designs/quic-design/quic-api.md

```diff
@@ -818,6 +876,340 @@ reason string in bytes.
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | C             |
+
+```c
+/*
+ * Used to control single stream operation. Calling this function determines the
+ * nature of the default stream which will automatically be created on the QUIC
+ * connection SSL object.
+ *
+ * The default mode is `SSL_DEFAULT_STREAM_MODE_AUTO_BIDI`.
+ *
+ * The modes are as follows:
+ *
+ *   - `SSL_DEFAULT_STREAM_MODE_NONE`: No default stream will ever be created.
+ *     The application is assumed to understand multi-stream operation.
+ *     Remotely-initiated streams are placed in the accept queue for application
+ *     consumption. `SSL_read` and `SSL_write` calls must be made on a QUIC
+ *     stream SSL object, not the QUIC connection SSL object, as no default
+ *     stream will be associated with it.
+ *
+ *   - `SSL_DEFAULT_STREAM_MODE_AUTO_BIDI`: "First stream wins" mode of
+ *     operation for single-stream usage. If `SSL_write` is called before the
+ *     peer opens a remotely-initiated stream, a locally-initiated bidirectional
+ *     stream is created and bound as the default stream. If the peer opens a
+ *     remotely-initiated stream before the local application calls `SSL_write`
+ *     (with `len > 0`) for the first time, that stream is bound as the default
+ *     stream, which may be bidirectional or unidirectional; if it is
+ *     unidirectional, calls to `SSL_write` will fail. Attempts to create
+ *     additional streams by the peer are automatically rejected unless
+ *     the application opts in (API TBD).
+ *
+ *   - `SSL_DEFAULT_STREAM_MODE_AUTO_UNI`: "First stream wins" mode of
+ *     operation for single-stream usage, with a unidirectional stream. This
+ *     functions identically to `SSL_DEFAULT_STREAM_MODE_AUTO_BIDI`, but if the
+ *     local application calls `SSL_write` prior to the peer creating a
+ *     remotely-initiated stream, a unidirectional TX-only stream is created and
+ *     bound as the default stream. Thereafter, calls to `SSL_read` will fail.
+ *     If the peer creates a remotely-initiated stream prior to the first call
+ *     to `SSL_write` (with `len > 0`), that stream will be bound as the default
+ *     stream; note that a bidirectional stream may be bound in this case.
+ *     Attempts to create additional streams by the peer are automatically
+ *     rejected unless the application opts in (API TBD).
+ *
+ * This function must be called before a default stream object is created, for
+ * example before initiating a connection. If the function is too late to have
+ * an effect, this function fails and returns 0. To switch to multi-stream
+ * operation after a default stream has been created, use `SSL_detach_stream`.
+ */
+#define SSL_DEFAULT_STREAM_MODE_NONE                0
+#define SSL_DEFAULT_STREAM_MODE_AUTO_BIDI           1
+#define SSL_DEFAULT_STREAM_MODE_AUTO_UNI            2
+
+__owur int SSL_set_default_stream_mode(SSL *ssl, uint32_t mode);
```

Comment:
This IMO overly complicates things for no reason. There should be only two modes - NONE for multi-stream operation and AUTO implying the BIDI stream being the default locally or remotely initiated based on whether SSL_write() or SSL_read() is called first.

That would also mean no SSL_attach_stream() and SSL_detach_stream() is necessary as multi-stream application simply sets the default mode to NONE if it wants to treat all streams the same way.

Reply:
It's probably true that `AUTO_UNI` will be quite a niche thing, but it's more a question of symmetry here. If the client does `SSL_read` first it's up to the server to initiate a remotely-initiated stream - and that could be bidirectional or unidirectional, that's not up to us. In this case whether `AUTO_BIDI` or `AUTO_UNI` is selected makes no difference, so even if we had just `AUTO`, we would have to handle the unidirectional case. So it feels like an odd asymmetry to not also support a locally-initiated unidirectional stream. But can be discussed.